File: 01.Searching/04. Exponential Search/BS_infiniteArray.js

let binarySearch = function (arr, x) {
    let start = 0, end = 1;
    // Iterate while start not meets end

    //setting the upper limit 
    while (x > arr[end]) {
        start = end;
        end = 2 * end;
        if (end >= arr.length) {
            end = arr.length - 1;
            break;
        }
    }
    //regular binary search.
    while (start <= end) {

        // Find the mid index
        let mid = Math.floor((start + end) / 2);
        // If element is present at mid, return True
        if (arr[mid] === x) return mid;
        // Else look in left or right half accordingly
        else if (arr[mid] < x)
            start = mid + 1;
        else
            end = mid - 1;
    }
    return -1;
}
// Driver code
let arr = [1, 3, 5, 7, 8, 9];
let x = 15;
console.log(binarySearch(arr, x));

File: 01.Searching/04. Exponential Search/BS_infiniteArray.java
public class BS_infiniteArray {
    static int bs(int[] arr, int target) {
        int low = 0, high = 1;

        // setting the upper limit aka high
        while (target >= arr[high]) {
            low = high; // optimization
            high = 2 * high;
            if (high >= arr.length - 1) {
                high = arr.length - 1;
                break;
            }
        }
        // regular binary Search.
        while (low <= high) {
            int mid = low + (high - low) / 2;
            if (arr[mid] == target)
                return mid;
            else if (arr[mid] < target) {
                low = mid + 1;
            } else
                high = mid - 1;
        }
        return -1;
    }

    public static void main(String[] args) {
        int arr[] = { 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11 };
        int target = 1001;
        System.out.println(bs(arr, target));
    }
}